zend_abstract_interface/zai_sapi/zai_sapi_ini.c

```diff
@@ -0,0 +1,43 @@
+#include "zai_sapi_ini.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+size_t zai_sapi_ini_entries_alloc(const char *src, char **dest) {
+    if (!src || !dest) return 0;
+
+    /* Prevent a potential dangling pointer in case the caller accidentally
+     * sent in an allocated dest.
+     */
+    if (*dest != NULL) return 0;
+
+    size_t len = strlen(src);
+
+    if ((*dest = (char *)malloc(len + 1)) == NULL) return 0;
+    memcpy(*dest, src, (len + 1));
+
+    return len;
+}
+
+void zai_sapi_ini_entries_free(char **entries) {
+    if (entries != NULL && *entries != NULL) {
+        free(*entries);
+        *entries = NULL;
+    }
+}
+
+size_t zai_sapi_ini_entries_realloc_append(char **entries, size_t entries_len, const char *key, const char *value) {
```

Comment:
https://github.com/DataDog/dd-trace-php/pull/1204/files#diff-1c20928ddec31938013846976fec494c8a394b606b2071ce82a80144fcdec40cR7 applies here as well

Reply:
Agreed - we need to return `-1` for errors here for consistency with `zai_sapi_ini_entries_alloc`.  👍  Although having an empty `key` or `value` wouldn't be valid in this case so those would also return `-1`. In this case, this function should never return `0`. I have tests for `NULL` key/values but not empty strings; I need to add those tests. 🙂